docs/hugo/content/design/ADR-2023-04-Patch-Collections.md

```diff
@@ -0,0 +1,115 @@
+
+In principle, an ASO custom resource (CR) should be the goal-state for the resource and should be a complete definition of the desired state. Whether the collection is empty or missing entirely should not matter, we should interpret that as "clear the collection".
+
+### Option 1: Optional Properties
+
+Tack an extra pointer onto every slice/map property across all generated types.
+
+Instead of:
+
+``` go
+MyProperty []Type 'json:"myProperty,omitempty"'
+```
+
+we would generate:
+
+``` go
+MyProperty *[]Type 'json:"myProperty,omitempty"'
+```
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* CON: Horrible to use in code because you must dereference the property to range over it (or really do much of anything with it).
+* CON: Breaking change for any consumers of our object model.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+
+### Option 2: Remove `omitempty` 
+
+Remove omitempty from slice/map properties across all generated types.
+
+``` go
+MyProperty []Type 'json:"myProperty"'. 
+```
+
+This means that the slice or map will always be serialized, even if empty.
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Not a breaking change for consumers of our object model.
+* CON: Empty collections will always be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: If we extend this to all properties, all properties will be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+* QUERY: It is possible that some Azure services treat `myProperty: null` as different than omitting the property entirely, although we do not actually know of any cases where that is true.
+
+### Option 3: Custom JSON Serialization
+
+Use a custom JSON serialization library that supports an `omitnil` equivalent, such as [jettison](https://github.com/wI2L/jettison).
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Not a breaking change for consumers of our object model.
+* CON: Breaks CRD YAML deserialization for the ASO controller (as the [sigs.k8s.io/yaml]( https://github.com/kubernetes-sigs/yaml) library we use works with standard JSON serialization attributes)
+* CON: Also breaks YAML serialization for `asoctl` (for the same reason)
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+
+### Option 4: Remove `omitempty` for selected ARM types only
+
+As for Option 2, above, we apply the change only selected ARM Spec types, leaving the core ASO Spec and Status types unmodified.
+
+Selection occurs through our existing `ObjectModelConfiguration` allowing us to explicitly target collections (and other properties) that we know are problematic.
```

Comment:
Do we do this at the property level or at the type level (or maybe at the resource level?)

I kinda think we shouldn't do it at the property level because it's _too_ fine-grained? AFAIK, when RPs have this problem they have it for the whole resource (including complex sub-properties, etc). Though that may be up to RPs individual implementations...

Basically wondering if something liek: "You can specify that a whole resource including its whole type hierarchy shouldn't use `omitempty`" and when we run into this for say, `storageAccount`, we just blat that onto the whole resource and call it good rather than try and be overly surgical?

Reply:
Or maybe we want the ability to do it at either the type or the property level? See my comment on #2916 about this